proposals/distribution.md

```diff
@@ -0,0 +1,156 @@
+
+The following entries should be added to the [scope table][scope]:
+
+* “Specifying authentication and authorization schemes”.
+    Docker's current registry uses an [extension][token] of the [`Bearer`][rfc6750] [auth scheme][rfc7235-s2.1].
+    Work on specifying Docker's scheme will continue independently, and is orthogonal to the registry API.
+
+    * What: Specifying authentication and authorization schemes
+    * In/Out/Future: Out of scope
+    * Status: N/A
+    * Description: Defining protocols for authenticating and authorizing distribution access.
+    * Why: As an HTTP-based protocol, clients and servers can negotiate authentication via HTTP's [challenge-response authentication framework][rfc7235-s2.1].
+        There is no need for the distribution specification to choose a particular authentication scheme, because clients receiving 401 and 407 responses can use IANA's [HTTP Authentication Scheme Registry][iana-auth] to look up referenced schemes and take appropriate action.
+
+* “Creating a reference spec for optional DNS based naming and discovery”.
+    Discovery and registry protocols are completely separate and do not need to be added together.
+    This entry replaces part of the previous “Creating Reference spec for optional DNS based naming & distribution” entry.
+
+    * What: Creating a reference spec for optional DNS based naming and discovery
+    * In/Out/Future: In scope for future specification
+    * Status: Work not yet started
+    * Description: Define a protocol for resolving an image name to retrieval information.
+        When we address this, we will also allow for alternative name-to-image discovery protocols in parallel with the OCI-specified protocol.
+    * Why: It is reasonable to provide a standardized way to use DNS based distribution in conjunction with OCI without requiring its use.
+        There are many good use cases for DNS based distribution, but not all use cases support this.
+        Furthermore, encoding the location of a bundle into the bundle can cause issues with downloads from alternate locations other than the origin specified in the name.
+
+* “Specifying a distribution method”.
+    This entry replaces part of the previous “Creating Reference spec for optional DNS based naming & distribution” and “Standardizing on a particular Distribution method” entries.
+
+    Retrieving image indexes covers the current “tag listing” (e.g. “what named manifests are in `library/busybox`?”), because tags are entries in the image format's [`manifests` array][manifests].
+    Other tag-listing endpoints needed for backwards-compatibility are therefore in scope as well.
+
+    Grouping image indexes in repositories is considered part of distribution policy or content management, which are out of scope for this entry's per-image action.
+    For example, “what images are under `library/`?” is out of scope for this project.
```

Comment:
From the perspective of this specification, there is no relationship between `library` and `library/foo`. This allows one to arbitrarily map the registry namespace to different schemas. In this specification, they are just treated as a name.

Reply:
I like this approach as well - having developed tools that separated some prefix / collection of a namespace (e.g., library) from the rest (e.g., "foo") led to much error, when really we can just call the whole thing a group (name or namespace?) It's a bit like the "file hierarchy" in Google Storage - it just looks like folders / files but it's really just an index with slashes in it. So I agree -
 instead of having to classify that "library" is a different thing from "library/foo" I think thus it's better to have them treated as a name, and then tools that want to mimic a filesystem (or search, etc.) can do things akin to name.startswith('library') etc.

Reply:
> From the perspective of this specification, there is no relationship between `library` and `library/foo`. This allows one to arbitrarily map the registry namespace to different schemas. In this specification, they are just treated as a name.

That's fine with me.  The point I'm trying to make around here is that if `library/foo` is representable as an image-index, then operations on higher-level objects like `library/` are out of scope for this PR.  There was a lot of wording debate on this in #44 which was eventually resolved in #46, but if you have different wording preferences for that idea I'm open to changes.

Reply:
@wking These are really technical considerations that I don't think belong in the proposal. The only takeaway here is that `_` start is reserved in the repo names to avoid overlap with endpoints like `/v2/_catalog` and to allow space for in-band backend mappings.

Reply:
Let me clarify: repository naming is completely in scope here. The API isn't very useful without it. @vsoch's understanding is correct. How repository names are mapped to image names is up to the implementation or defined by convention.

Reply:
> These are really technical considerations that I don't think belong in the proposal.

I want to set clear enough scope boundaries that operations like [the current `/v2/_catalog`][1] or the API form of [user's image-index listing][2], while keeping the API form of [tags within an image-index][3] in scope.  Since there's already a `/v2/_catalog` in Docker's `api.md`, I think we need to get technical enough to make that distinction.

> The only takeaway here is that `_` start is reserved in the repo names to avoid overlap with endpoints like `/v2/_catalog` and to allow space for in-band backend mappings.

I'm fine with the OCI distribution API to reserve a namespace like `/v2/_.*` for backend extentions.  I don't think that conflicts with leaving `/v2/_catalog` as a docker/distribution-specific extension.

> Let me clarify: repository naming is completely in scope here. The API isn't very useful without it.

That's fine with me.  What does it have to do with making “[what images are under `library/`?][2]” out of scope for the distribution API?

[1]: https://github.com/docker/distribution/blob/5cb406d511b7b9163bff9b6439072e4892e5ae3b/docs/spec/api.md#catalog
[2]: https://hub.docker.com/u/library/
[3]: https://hub.docker.com/r/library/alpine/tags/

Reply:
The thing you linked in the hub is not, at all, an implementation of this specification. The catalog endpoint is just a raw listing of repositories. It's out of scope because it is a management function of the registry and not required for runtimes to pull an image.

Reply:
@wking Can you fix the wording here? This should only call out that hierarchies aren't a part of this specification.

Reply:
This has been fixed under #50.